### PR TITLE
fix(server): use compression for non big values

### PR DESCRIPTION
**Pull request by adiholden**
Changes in this PR:
1. Fix bug in snapshot serialization that caused us not to do compression at all
2. Enable compression if big value serialization is enabled (flag max_serialization_chunk_size !=0)
3. Change dragonfly rand string lua call to generate more compressible string.
4. Add statistics to compression and check stats in tests
5. Simplify the seeder for generating big values. Instead of defining keys as keys of big value, define the total number of entries of big values when running seeder.

### CODE REVIEW

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -1644,6 +1638,8 @@ void SerializerBase::CompressBlob() {
   memcpy(dest.data(), compressed_blob.data(), compressed_blob.length());
   mem_buf_.CommitWrite(compressed_blob.length());
   ++compression_stats_->compressed_blobs;
+  auto& stats = ServerState::tlocal()->stats;
```
Comment (romange) [inline src/server/rdb_save.cc:1641]: nit: ugh, we already have `compression_stats_` that go nowhere and now we also have this. Shoud we, maybe, expose `compression_stats_` via API and then record the stats in the upper layer?

File: src/server/server_family.cc [hunk 2]
```diff
@@ -2373,6 +2373,7 @@ void ServerFamily::Info(CmdArgList args, const CommandContext& cmd_cntx) {
     append("rdb_save_usec", m.coordinator_stats.rdb_save_usec);
     append("rdb_save_count", m.coordinator_stats.rdb_save_count);
     append("big_value_preemptions", m.coordinator_stats.big_value_preemptions);
+    append("compressed_blobs", m.coordinator_stats.compressed_blobs);
```
Comment (romange) [inline src/server/server_family.cc:2376]: i think it belongs to "PERSISTENCE"  section below.

File: tests/dragonfly/instance.py [hunk 3]
```diff
@@ -416,7 +416,7 @@ def create(self, existing_port=None, path=None, version=100, **kwargs) -> DflyIn
         args.setdefault("noversion_check", None)
         # MacOs does not set it automatically, so we need to set it manually
         args.setdefault("maxmemory", "8G")
-        vmod = "dragonfly_connection=1,accept_server=1,listener_interface=1,main_service=1,rdb_save=1,replica=1,cluster_family=1,proactor_pool=1,dflycmd=1"
+        vmod = "dragonfly_connection=1,accept_server=1,listener_interface=1,main_service=1,rdb_save=2,replica=1,cluster_family=1,proactor_pool=1,dflycmd=1,snapshot=2"
```
Comment (romange) [inline tests/dragonfly/instance.py:419]: are you sure you want to keep it? it will produce rather big logs.

File: tests/dragonfly/replication_test.py [hunk 4]
```diff
@@ -49,14 +49,16 @@ async def wait_for_replicas_state(*clients, state="online", node_role="slave", t
         # Quick general test that replication is working
         (1, 3 * [1], dict(key_target=1_000), 500),
         # A lot of huge values
-        (2, 2 * [1], dict(key_target=1_000, huge_value_percentage=2), 500),
+        (2, 2 * [1], dict(key_target=1_000, huge_value_target=30), 500),
```
Comment (romange) [inline tests/dragonfly/replication_test.py:52]: same - does it belong to this PR?

File: tests/dragonfly/replication_test.py [hunk 5]
```diff
@@ -2320,7 +2323,7 @@ async def test_master_stalled_disconnect(df_factory: DflyInstanceFactory):
     c_master = master.client()
     c_replica = replica.client()
 
-    await c_master.execute_command("debug", "populate", "200000", "foo", "500")
+    await c_master.execute_command("debug", "populate", "200000", "foo", "500", "RAND")
```
Comment (romange) [inline tests/dragonfly/replication_test.py:2326]: why did you change these tests?

**Review by romange — COMMENTED** [review]
I do not understand the changes in tests. Are these needed?

File: tests/dragonfly/replication_test.py (hunk 5 above)
Comment (adiholden) [reply tests/dragonfly/replication_test.py:2326]: because we use compression now the debug populate is very compressed unless using the RAND option and therefore the test would fail as the master will send all the data to socket and wont get blocked. We need to get blocked on write to socket and we need to have enough data to send to be blocked

File: tests/dragonfly/instance.py (hunk 3 above)
Comment (adiholden) [reply tests/dragonfly/instance.py:419]: I will revert this

File: src/core/interpreter.cc [hunk 6]
```diff
@@ -386,13 +386,26 @@ int DragonflyRandstrCommand(lua_State* state) {
...
+
   auto push_str = [dsize, state, &buf]() {
     static const char alphanum[] =
         "0123456789"
         "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
         "abcdefghijklmnopqrstuvwxyz";
-    for (int i = 0; i < dsize; ++i)
-      buf[i] = alphanum[rand() % (sizeof(alphanum) - 1)];
+
+    static const char pattern[] = "DRAGONFLY";
```
[the remark this answers is not in the document]
Comment (adiholden) [reply src/core/interpreter.cc:396]: we enable compression, but if I want to check the compression stats in replication tests I need this random string which is used in this tests to be compressable

File: src/server/server_family.cc (hunk 2 above)
Comment (adiholden) [reply src/server/server_family.cc:2376]: actually not sure it belongs to   "PERSISTENCE" as all the stats there are related for current/last save operations. but this can be related to replication as well.

File: tests/dragonfly/replication_test.py (hunk 4 above)
Comment (adiholden) [reply tests/dragonfly/replication_test.py:52]: I changed the logic of the seeder for huge values
Now we have by default huge values in seeder, which are the once who makes us trigger the big value serialization flow.
The problem with huge_value_percentage is that if we have large key_target this will lead also to large target of huge values and we can get to OOM fast in this case.
Before the flow was enabled because the value of serialization_max_chunk_size was kind of small leading us to test the big value serialization flow but it caused the regular flow of SliceSnapshot::PushSerialized to be disabled as we always flush before.

File: src/server/rdb_save.cc (hunk 1 above)
Comment (adiholden) [reply src/server/rdb_save.cc:1641]: are you suggesting something like `debug compression` ?
I can do that, but I suggest to push it in a new PR.
If you dont have more comments on this one we can merge it and you can pull this to your change with lz4 to test and I can work at the same time on this new api

**Review by romange — APPROVED** [review]
